Add physical DB to CDK, and connect lambda

jonathonherbert opened the pull request:
## What does this change?

Adds an RDS to the content-audit stack, and wires it up to the playwright-runner lambda.

In addition to that, we also add an ASG to the stack, defaulting to size 0, that hosts a bastion instance. The instance has a script to scale in the ASG when an SSH connection has not been detected for the last 2.5hrs (borrowed from Pinboard, thanks @twrichards!)

We use RDS Proxy to address a problem common when scaling out serverless functions that need database connections — Prisma has a great writeup [here](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections#the-serverless-challenge). RDS also enables IAM authentication for database connections, which means we can generate short-lived credentials for database interactions, sidestepping problems like credential rotation.

The architecture looks like: 

```mermaid
flowchart LR
  L[Lambda] --IAM auth--> P[RDS Proxy] --> RDS[RDS instance]
  LO[Local machine] --> B[EC2 instance] --IAM auth--> P
```

This adds complexity in that we need some scripting to make tunnelling to the database, and authenticating, trivial for local development. To that end, this PR includes three commands within the `playwright-runner` project, `write-local-env`, `rds-tunnel` and `stop-rds-tunnel`:
- `write-local-env` writes a `.env` file with a connection string valid for the local DB.
- `rds-tunnel` creates a tunnel to the bastion instance, generates an IAM credential for the database, and writes a postgres connection string for that environment to the local `.env` file, for use by Prisma or manually in a DB client.
- `stop-rds-tunnel` rewrites the `.env` with the local environment again, and kills the tunnel process to free up the port.

## Why are there so many lines?

Don't worry — most of them are generated by Prisma, which got an upgrade from 6 to 7 as a part of this work. Unless I've added GitHub comments, tou can safely ignore the files in the `prisma` folder.

## How to test

- The stack should deploy correctly — the project is `Content Platforms::content-audit`. There's only a CODE stack at present!
- Try scaling out the [ASG](https://eu-west-1.console.aws.amazon.com/ec2/home?region=eu-west-1#AutoScalingGroupDetails:id=content-audit-bastion-CODE;view=details) by 1 (it may already be scaled out!), and establishing a local connection to the database with SSM with `npm run rds-tunnel`. You should be able to successfully connect to the database with a DB client of your choice using the connection string written to `.env`.
- Try running [the lambda](https://eu-west-1.console.aws.amazon.com/lambda/home?region=eu-west-1#/functions/playwright-runner?subtab=envVars&tab=testing) with a test payload like the one below. It should succeed. The message `Current run count: 0 (should be 0!)` indicates we've successfully connected to the DB.
```
{
  "body": "{\"url\": \"https://theguardian.com\"}"
}
```

<img width="1399" height="264" alt="Screenshot 2026-01-26 at 16 14 18" src="https://github.com/user-attachments/assets/b4c6769c-4305-4b91-b1e3-32f6a83e2544" />

- [x] Tested all the above in CODE.

jonathonherbert committed:
Add RDS to CDK

jonathonherbert committed:
Add connection URL to lambda, and hook up lambda to DB

jonathonherbert committed:
Correct DB name

jonathonherbert committed:
Add DB proxy to enable lambda to reuse connections

jonathonherbert committed:
Add bastion ASG

jonathonherbert committed:
Modify local docker configuration to correctly access DB

jonathonherbert committed:
Tweaks to test DB connection in AWS

jonathonherbert committed:
Add startup script (borrowed from Pinboard) to DB bastion

jonathonherbert committed:
Update AMI to new bastion type

jonathonherbert committed:
Temporarily remove the lambda to spin up the ECR as part of the template

jonathonherbert committed:
Debugging DB connection string

jonathonherbert committed:
Check in migrations

jonathonherbert committed:
Add install and shebang to setup script

jonathonherbert committed:
Minor tidy of entrypoint script

twrichards commented:
You could also borrow the scripts that scale up the ASG wait for it to become usable then create the tunnel etc.

jonathonherbert committed:
Bump node to 22.21.1

jonathonherbert committed:
Refactor application to use IAM signer to acquire database password

jonathonherbert committed:
Factor db code to accept password for test env

jonathonherbert committed:
Add missing security group connections for dbProxy, bastion and lambda

grantConnect is not enough

jonathonherbert committed:
Add scripts for local tunnelling

find rds proxy endpoint, tunnel, get credentials and print to stdout

jonathonherbert committed:
Add script to run locally via tunnel

jonathonherbert committed:
Bump to prisma 7, in an attempt to fix migration errors

(or provide a clearer message than User was denied access on the database (not available))

jonathonherbert committed:
URI-encode password as part of connection string

jonathonherbert committed:
Align testing environment with changes

jonathonherbert committed:
Force exit after tests are done

Although hopefully I'll get to the bottom of what's holding on to resources that's forcing the app to hang

jonathonherbert committed:
Reinstate password in env var for local running

jonathonherbert committed:
Log credentials to debug

jonathonherbert committed:
uriencode password

jonathonherbert committed:
Remove configuration params

jonathonherbert committed:
Add root certificate to Docker image, and reference in connection string

jonathonherbert committed:
Tidy up, renaming DB to contentaudit for consistency across environments (AWS will not allow hyphens in DB names)

jonathonherbert commented on .github/workflows/build.yml:
```diff
@@ -118,3 +119,4 @@ jobs:
             cdk.out:
               - artifacts/cdk.out/ContentAudit-euwest-1-CODE.template.json
               - artifacts/cdk.out/ContentAudit-euwest-1-PROD.template.json
+              - cdk/startup.sh
```
The startup script for the bastion host (the ephemeral EC2 instance that we can use to connect to the remote DB locally)

jonathonherbert committed:
Remove unnecessary DOCKER_OPTS

as this was not the source of the RDS connection issues

jonathonherbert commented on projects/page-runner/prisma/schema.prisma:
```diff
@@ -1,11 +1,10 @@
 generator client {
-  provider = "prisma-client-js"
+  provider = "prisma-client"
```
This PR migrates to the latest version of Prisma, which is why there are a great many changes to the client, as well as these changes to the configuration.

jonathonherbert commented on projects/page-runner/docker-compose.yml:
```diff
@@ -5,14 +5,14 @@ services:
-            - POSTGRES_USER=content-audit
-            - POSTGRES_PASSWORD=content-audit
-            - POSTGRES_DB=content-audit
+            - POSTGRES_USER=contentaudit
```
AWS doesn't permit hyphens in database names, so for consistency across environments, I've changed this.

jonathonherbert commented on projects/page-runner/entrypoint.sh:
```diff
@@ -2,13 +2,13 @@
 else
   printf "Running in lambda on node $NODE_VERSION\n"
-  exec /usr/bin/npx aws-lambda-ric $1
+  exec /usr/bin/npx aws-lambda-ric "$1"
```
This is just linting — quotes are good practice in shell substitions.

jonathonherbert commented on projects/playwright-runner/package.json:
```diff
@@ -5,24 +5,29 @@
     "migrate": "prisma migrate deploy",
-    "postinstall": "prisma generate"
+    "postinstall": "prisma generate",
+    "rds-tunnel": "sh -c 'ssm ssh -t content-audit,CODE --profile capi --tunnel 7658:$(node ./scripts/get-rds-proxy-endpoint.ts --stage CODE):5432' && echo \"\npassword:\n\" && node ./scripts/generate-rds-password.ts",
```
A handy command to tunnel to the bastion EC2 instance, and generate a password to connect.

jonathonherbert committed:
More tidies after a review

jonathonherbert committed:
Add build number to cdk, and pass in from CI, to link deployed infra with the created Docker image

jonathonherbert committed:
Move ECR out of stack, and pass in lookup values via SSM params

jonathonherbert committed:
Add infra stack, and house ECR repo in that

jonathonherbert committed:
playwright-runner -> page-runner

jonathonherbert committed:
Move GitHub ECR role into infra stack

jonathonherbert committed:
Add infra stack to riffraff action

jonathonherbert committed:
Adjust output of tunnel command, which no longer needs to output the password

jonathonherbert committed:
Correct name of infra app

jonathonherbert committed:
Instrument setup file to understand what's taking so long

jonathonherbert committed:
Remote infra stack from riff-raff deployment, for now

RikRootsGuardian reviewed (approved):
Happy to approve following pair session
